### ISSUE TITLE
multi-thread issue

### ISSUE THREAD

**Issue by ldy26098** [RuedigerMoeller/fast-serialization#70]
first of all ,thanks for the effort on fst.

my server used kafka for message queue , and java built-in serialization for decoder/encoder . 

yesterday i changed decoder/encoder to FST default configuration and it works well . today i tested in production environment . when i have 2000 clients connected , the server works fine. when i have 3000 clients , kafka cannot receive message. then i changed decoder/encoder back to java built-in , and 3000 clients is ok.

i tried several times , but it did have a problem in high pressure with FST. any suggestion for my case?

my code just like below:

    public static IMessage bytes2msg(byte[] bytes) {
        return MQUtils.getFSTConf().asObject(bytes);
    }

    public byte[] toBytes(IMessage msg) {
			return MQUtils.getFSTConf().asByteArray(msg);
		}

    public class MQUtils {

	private static FSTConfiguration conf = FSTConfiguration.createDefaultConfiguration();

        public static FSTConfiguration getFSTConf() {
		return conf;
	}
    }

**Comment by RuedigerMoeller** [comment]
What might happen is contention on FSTConfiguration (too much threads trying to access it concurrently). You can workaround this by using a ThreadLocal instead of a static singleton FSTConfiguration. 
There is a downside on this, because Permspace migth overflow up as FSTConfiguration caches generated reflection accessors. If you work with >hundred threads you might need to increase Permspace then.
A more refined variant is to share like 8 FSTConfigurations via ThreadLocal trickery (e.g. use Thread object id % 8 to choose one of 8 static FSTConfigurations) but not have one instance for each thread.

Ofc there might be arbitrary other reasons e.g. as FST is significantly faster, race conditions might show up which were hidden when using slower serialization.

I know of a system using FSTConfiguration in high a pressure scenario, however its more like 100 clients with high per-client load, so I would assume there is no general issue under pressure.

**Comment by RuedigerMoeller** [comment]
Example to reduce contention without memory bloat:

```java
    public static FSTConfiguration[] configurations = new FSTConfiguration[8];
    public static ThreadLocal<FSTConfiguration> conf = new ThreadLocal<FSTConfiguration>() {
        @Override
        protected FSTConfiguration initialValue() {
            return configurations[((int) (Thread.currentThread().getId() % 8))];
        }
    };
    static {
        for (int i = 0; i < configurations.length; i++) {
            configurations[i] = FSTConfiguration.createDefaultConfiguration();
            LRUtil.registerDefaultClassesFST(FST2ByteCompressed.conf);
        }
    }
```

**Comment by ldy26098** [comment]
thanks for the reply .

i will keep on testing .

**Comment by RuedigerMoeller** [comment]
Just made scaling tests with kontraktor3 with a per client FSTConfiguration and ran into a wall at ~2400 clients. Investigation showed, that metaspace (=java 8's PermSpace) grew massively caused by FST's caching of reflection-generated classes. This did not lead to exception but slowed down connection of new clients to the point it failed.

I don't know if this is your issue, however from the numbers you mention (2000 fine 3000 not) it might also be the reason your app does not scale up (though the fix above should solve it).

The solution is to share common reflection artifacts (constructors, fields) across FSTConfigurations (this is safe as they are immutable per classloader). I wrongly assumed the JVM will do this anyway ...

With this fix I could scale up to 20.000 long poll connections on a single worker thread (Nio).

delivered with 2.30
